src/rgw/driver/posix/rgw_sal_posix.cc

```diff
@@ -1178,39 +1184,145 @@ int POSIXObject::close(const DoutPrefixProvider* dpp)
 int POSIXObject::read(int64_t ofs, int64_t end, bufferlist& bl,
 		      const DoutPrefixProvider* dpp, optional_yield y)
 {
-  int64_t len = std::min(end - ofs, READ_SIZE);
+  int64_t len = std::min(end - ofs + 1, READ_SIZE);
+  ssize_t ret;
+
+  ret = lseek(obj_fd, ofs, SEEK_SET);
+  if (ret < 0) {
+    ret = errno;
+    ldpp_dout(dpp, 0) << "ERROR: could not seek object " << get_name() << " to "
+      << ofs << " :" << cpp_strerror(ret) << dendl;
+    return -ret;
+  }
+
+  ret = ::read(obj_fd, read_buf, len);
+  if (ret < 0) {
+    ret = errno;
+    ldpp_dout(dpp, 0) << "ERROR: could not read object " << get_name() << ": "
+                  << cpp_strerror(ret) << dendl;
+    return -ret;
+  }
+
+  bl.append(read_buf, ret);
+
+  return ret;
+}
+
+int POSIXObject::write_attr(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key, bufferlist& value)
+{
+  int ret;
+  std::string attrname;
+
+  ret = open(dpp);
+  if (ret < 0) {
+    return ret;
+  }
+
+  attrname = ATTR_PREFIX + key;
 
-  return ::read(obj_fd, read_buf, len);
+  ret = fsetxattr(obj_fd, attrname.c_str(), value.c_str(), value.length(), 0);
+  if (ret < 0) {
+    ret = errno;
+    ldpp_dout(dpp, 0) << "ERROR: could not write attribute " << attrname << " for object " << get_name() << ": " << cpp_strerror(ret) << dendl;
+    return -ret;
+  }
+
+  return 0;
 }
 
 int POSIXObject::POSIXReadOp::prepare(optional_yield y, const DoutPrefixProvider* dpp)
 {
-  return source->open(dpp);
+  int ret = source->open(dpp);
+  if (ret < 0)
+    return ret;
+
+  ret = source->stat(dpp);
+  if (ret < 0)
+    return ret;
+
+  ret = source->get_obj_attrs(y, dpp);
+  if (ret < 0)
+    return ret;
+
+  auto iter = source->get_attrs().find(RGW_ATTR_ETAG);
+  if (iter == source->get_attrs().end()) {
+    /* Sideloaded file.  Generate necessary attributes. Only done once. */
+    int ret = source->generate_attrs(dpp, y);
```

Comment:
this seems like a bad idea in general, as the files could be arbitrarily large. if the S3 API requires us to return *some* value for ETag, is there a fake string we could return instead? either something like "NOT-A-REAL-ETAG", or if clients expect it to look like an MD5 hash, then return the hash of an empty buffer?

Reply:
HTTP only requires some functional properties of etag--i.e., it's relationship to cache-control, I believe.  S3 I think actually requires it to be the MD5 hash of the byte stream in the object.  If it's the md5sum of an empty buffer, the client can complain about things, and I think boto3 at least does, for what it is worth.

Reply:
This is temporary, just to get s3-cmd to stop complaining.  In fact, much of this is temporary, as part of the PoC build.

Reply:
And also to test reading/writing xattrs